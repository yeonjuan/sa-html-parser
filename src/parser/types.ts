export enum InsertionMode {
  initialMode,
  afterInitialMode,
  // beforeHtmlMode,
  // beforeHeadMode,
  // inHeadMode,
  // inHeadNoScriptMode,
  // afterHeadMode,
  // inBodyMode,
  // textMode,
  // inTableMode,
  // inTableTextMode,
  // inCaptionMode,
  // inColumnGroupMode,
  // inTableBodyMode,
  // inRowMode,
  // inCellMode,
  // inSelectMode,
  // inSelectInTableMode,
  // inTemplateMode,
  // afterBodyMode,
  // inFramesetMode,
  // afterFramesetMode,
  // afterAfterBodyMode,
  // afterAfterFramesetMode,
}
