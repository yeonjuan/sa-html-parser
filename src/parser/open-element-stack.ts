export class OpenElementStack {
  public readonly elements: any[] = [];
  public stackTop = -1;
  public top: any = null;

  public push(element: any) {
    this.top = element;
    this.stackTop++;
    this.elements[this.stackTop] = element;
  }

  public pop() {
    this.stackTop--;
    this.top = this.elements[this.stackTop];
  }

  public popUntilElementPopped(element: any): any[] {
    let elements: any[] = [];
    while (this.stackTop > -1) {
      const poppedElement = this.top;
      elements.push(poppedElement);
      this.pop();
      if (poppedElement === element) {
        break;
      }
    }
    return elements;
  }
}
