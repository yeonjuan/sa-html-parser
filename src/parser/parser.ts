import {
  CommentNode,
  DoctypeNode,
  RootNode,
  AnyNode,
  ElementNode,
  ClosingElementNode,
  TextNode,
} from "../nodes";
import { Tokenizer } from "../tokenizer/tokenizer";
import {
  AnyHtmlToken,
  AttributeToken,
  CharacterLikeToken,
  CommentToken,
  DoctypeToken,
  EndTagToken,
  EofToken,
  HtmlTokenType,
  NullToken,
  StartTagToken,
} from "../tokens";
import { OpenElementStack } from "./open-element-stack";
import * as utils from "../common/utils";

export class Parser {
  private root!: RootNode;
  private html!: string;
  private tokenizer!: Tokenizer;
  private openElementStack = new OpenElementStack();
  private running = true;

  constructor() {}

  public parse(html: string): RootNode {
    this.html = html;
    this.tokenizer = Tokenizer.create(this.html);
    this.createRoot();

    while (this.running) {
      const token = this.tokenizer.getNextToken();
      if (!token) {
        break;
      }
      this.process(token);
    }

    return this.root;
  }

  private createRoot() {
    this.root = new RootNode();
    this.openElementStack.push(this.root);
  }

  private insertCommentNodeToRoot(node: CommentNode) {
    this.insertToCurrent(node);
    this.root.comments.push(node);
  }

  private insertToCurrent(node: AnyNode) {
    this.openElementStack.top.children.push(node);
  }

  private pushToOpenStack(node: any) {
    this.openElementStack.push(node);
  }

  private popFromOpenStackUntilTagName(tagName: string): null | any[] {
    let unclosedElements: null | any[] = null;
    for (let i = this.openElementStack.stackTop; i >= 0; i--) {
      const element = this.openElementStack.elements[i];
      if (element.tagName === tagName) {
        unclosedElements = this.openElementStack.popUntilElementPopped(element);
        break;
      }
    }
    return unclosedElements;
  }

  private process(token: AnyHtmlToken) {
    this[token.type]?.(token as any);
  }

  private [HtmlTokenType.Comment](token: CommentToken) {
    const comment = CommentNode.fromToken(token);
    this.insertToCurrent(comment);
    this.insertCommentNodeToRoot(comment);
  }

  private [HtmlTokenType.Doctype](token: DoctypeToken) {
    this.insertToCurrent(DoctypeNode.fromToken(token));
  }

  private [HtmlTokenType.StartTag](token: StartTagToken) {
    const tagNode = ElementNode.fromToken(token);
    this.insertToCurrent(tagNode);
    if (token.selfClosing) {
      tagNode.selfClosing = true;
    } else {
      this.pushToOpenStack(tagNode);
    }
  }

  private [HtmlTokenType.EndTag](token: EndTagToken) {
    const closing = ClosingElementNode.fromToken(token);
    const poppedElements = this.popFromOpenStackUntilTagName(
      token.tagName.value
    );
    if (!poppedElements) {
      throw new Error();
    }
    const element = utils.last<ElementNode>(poppedElements)!;
    element.children = utils.getChildrenRecursively(element);
    element.closingElement = closing;
    element.end = closing.end;
    element.loc.end = closing.loc.end;
    element.range[1] = closing.range[1];
  }

  private [HtmlTokenType.CharacterLike](token: CharacterLikeToken) {
    if (this.openElementStack.top.children?.length) {
      const lastChild: any = utils.last(this.openElementStack.top.children);
      if (lastChild.type === "Text") {
        const textNode = lastChild as TextNode;
        textNode.value += token.value.value;
        textNode.end = token.end;
        textNode.loc.end = token.value.loc.end;
        return;
      }
    }
    this.insertToCurrent(TextNode.fromToken(token));
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.EOF](token: EofToken) {
    const poppedElements = this.openElementStack.popUntilBeforeElementPopped(
      this.root
    );
    if (poppedElements.length) {
      for (let i = poppedElements.length - 2; i >= 0; i--) {
        this.root.children.push(poppedElements[i]);
      }
    }
    this.running = false;
    const lastChild = utils.last<AnyNode>(this.root.children);
    this.root.range[1] = lastChild?.range[1] || 0;
    this.root.end = lastChild?.range[1] || 0;
    this.root.loc.end = lastChild?.loc.end || {
      column: 0,
      line: 1,
    };
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.Null](token: NullToken) {}
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private [HtmlTokenType.Attribute](token: AttributeToken) {}
}
