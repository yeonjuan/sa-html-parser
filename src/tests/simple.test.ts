import { parse } from "../parser";
import * as fs from "fs";
import * as path from "path";

describe("large", () => {
  const simpleHTML = fs.readFileSync(
    path.join(__dirname, "simple-html"),
    "utf-8"
  );

  it("simple.html", () => {
    const result = parse(simpleHTML);
    expect(result).toMatchSnapshot();
  });
});
