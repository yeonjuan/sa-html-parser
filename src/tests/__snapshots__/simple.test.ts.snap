// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`large simple.html 1`] = `
RootNode {
  "children": Array [
    DoctypeNode {
      "end": 15,
      "loc": Object {
        "end": Object {
          "column": 15,
          "line": 1,
        },
        "start": Object {
          "column": 0,
          "line": 1,
        },
      },
      "range": Array [
        0,
        15,
      ],
      "start": 0,
      "type": "DocumentType",
    },
    TextNode {
      "end": 16,
      "loc": Object {
        "end": Object {
          "column": 16,
          "line": 1,
        },
        "start": Object {
          "column": 15,
          "line": 1,
        },
      },
      "range": Array [
        15,
        16,
      ],
      "start": 15,
      "type": "Text",
      "value": "
",
    },
    ElementNode {
      "children": Array [
        TextNode {
          "end": 25,
          "loc": Object {
            "end": Object {
              "column": 2,
              "line": 3,
            },
            "start": Object {
              "column": 6,
              "line": 2,
            },
          },
          "range": Array [
            22,
            25,
          ],
          "start": 22,
          "type": "Text",
          "value": "
  ",
        },
        ElementNode {
          "children": Array [
            TextNode {
              "end": 36,
              "loc": Object {
                "end": Object {
                  "column": 4,
                  "line": 4,
                },
                "start": Object {
                  "column": 8,
                  "line": 3,
                },
              },
              "range": Array [
                31,
                36,
              ],
              "start": 31,
              "type": "Text",
              "value": "
    ",
            },
            CommentNode {
              "end": 48,
              "loc": Object {
                "end": Object {
                  "column": 16,
                  "line": 4,
                },
                "start": Object {
                  "column": 4,
                  "line": 4,
                },
              },
              "range": Array [
                36,
                48,
              ],
              "start": 36,
              "type": "Comment",
              "value": " foo ",
            },
            CommentNode {
              "end": 48,
              "loc": Object {
                "end": Object {
                  "column": 16,
                  "line": 4,
                },
                "start": Object {
                  "column": 4,
                  "line": 4,
                },
              },
              "range": Array [
                36,
                48,
              ],
              "start": 36,
              "type": "Comment",
              "value": " foo ",
            },
            TextNode {
              "end": 53,
              "loc": Object {
                "end": Object {
                  "column": 4,
                  "line": 5,
                },
                "start": Object {
                  "column": 16,
                  "line": 4,
                },
              },
              "range": Array [
                48,
                53,
              ],
              "start": 48,
              "type": "Text",
              "value": "
    ",
            },
            ElementNode {
              "children": Array [
                TextNode {
                  "end": 61,
                  "loc": Object {
                    "end": Object {
                      "column": 12,
                      "line": 5,
                    },
                    "start": Object {
                      "column": 9,
                      "line": 5,
                    },
                  },
                  "range": Array [
                    58,
                    61,
                  ],
                  "start": 58,
                  "type": "Text",
                  "value": "bar",
                },
              ],
              "closingElement": ClosingElementNode {
                "end": 67,
                "loc": Object {
                  "end": Object {
                    "column": 18,
                    "line": 5,
                  },
                  "start": Object {
                    "column": 12,
                    "line": 5,
                  },
                },
                "range": Array [
                  61,
                  67,
                ],
                "start": 61,
                "type": "ClosingElement",
              },
              "end": 67,
              "loc": Object {
                "end": Object {
                  "column": 18,
                  "line": 5,
                },
                "start": Object {
                  "column": 4,
                  "line": 5,
                },
              },
              "openingElement": OpeningElementNode {
                "attributes": Array [],
                "end": 58,
                "loc": Object {
                  "end": Object {
                    "column": 9,
                    "line": 5,
                  },
                  "start": Object {
                    "column": 4,
                    "line": 5,
                  },
                },
                "range": Array [
                  53,
                  58,
                ],
                "start": 53,
                "type": "OpeningElement",
              },
              "range": Array [
                53,
                67,
              ],
              "selfClosing": false,
              "start": 53,
              "tagName": "div",
              "type": "Element",
            },
            TextNode {
              "end": 72,
              "loc": Object {
                "end": Object {
                  "column": 4,
                  "line": 6,
                },
                "start": Object {
                  "column": 18,
                  "line": 5,
                },
              },
              "range": Array [
                67,
                72,
              ],
              "start": 67,
              "type": "Text",
              "value": "
    ",
            },
            ElementNode {
              "children": Array [],
              "closingElement": null,
              "end": 90,
              "loc": Object {
                "end": Object {
                  "column": 22,
                  "line": 6,
                },
                "start": Object {
                  "column": 4,
                  "line": 6,
                },
              },
              "openingElement": OpeningElementNode {
                "attributes": Array [
                  AttributeNode {
                    "end": 87,
                    "loc": Object {
                      "end": Object {
                        "column": 19,
                        "line": 6,
                      },
                      "start": Object {
                        "column": 11,
                        "line": 6,
                      },
                    },
                    "name": AttributeNameNode {
                      "end": 81,
                      "loc": Object {
                        "end": Object {
                          "column": 13,
                          "line": 6,
                        },
                        "start": Object {
                          "column": 11,
                          "line": 6,
                        },
                      },
                      "range": Array [
                        79,
                        81,
                      ],
                      "start": 79,
                      "type": "AttributeName",
                      "value": "id",
                    },
                    "range": Array [
                      79,
                      87,
                    ],
                    "start": 79,
                    "type": "Attribute",
                    "value": AttributeValueNode {
                      "end": 87,
                      "loc": Object {
                        "end": Object {
                          "column": 19,
                          "line": 6,
                        },
                        "start": Object {
                          "column": 14,
                          "line": 6,
                        },
                      },
                      "range": Array [
                        82,
                        87,
                      ],
                      "start": 82,
                      "type": "AttributeValue",
                      "value": "\\"baz\\"",
                    },
                  },
                ],
                "end": 90,
                "loc": Object {
                  "end": Object {
                    "column": 22,
                    "line": 6,
                  },
                  "start": Object {
                    "column": 4,
                    "line": 6,
                  },
                },
                "range": Array [
                  72,
                  90,
                ],
                "start": 72,
                "type": "OpeningElement",
              },
              "range": Array [
                72,
                90,
              ],
              "selfClosing": true,
              "start": 72,
              "tagName": "input",
              "type": "Element",
            },
            TextNode {
              "end": 93,
              "loc": Object {
                "end": Object {
                  "column": 2,
                  "line": 7,
                },
                "start": Object {
                  "column": 22,
                  "line": 6,
                },
              },
              "range": Array [
                90,
                93,
              ],
              "start": 90,
              "type": "Text",
              "value": "
  ",
            },
          ],
          "closingElement": ClosingElementNode {
            "end": 100,
            "loc": Object {
              "end": Object {
                "column": 9,
                "line": 7,
              },
              "start": Object {
                "column": 2,
                "line": 7,
              },
            },
            "range": Array [
              93,
              100,
            ],
            "start": 93,
            "type": "ClosingElement",
          },
          "end": 100,
          "loc": Object {
            "end": Object {
              "column": 9,
              "line": 7,
            },
            "start": Object {
              "column": 2,
              "line": 3,
            },
          },
          "openingElement": OpeningElementNode {
            "attributes": Array [],
            "end": 31,
            "loc": Object {
              "end": Object {
                "column": 8,
                "line": 3,
              },
              "start": Object {
                "column": 2,
                "line": 3,
              },
            },
            "range": Array [
              25,
              31,
            ],
            "start": 25,
            "type": "OpeningElement",
          },
          "range": Array [
            25,
            100,
          ],
          "selfClosing": false,
          "start": 25,
          "tagName": "body",
          "type": "Element",
        },
        TextNode {
          "end": 101,
          "loc": Object {
            "end": Object {
              "column": 10,
              "line": 7,
            },
            "start": Object {
              "column": 9,
              "line": 7,
            },
          },
          "range": Array [
            100,
            101,
          ],
          "start": 100,
          "type": "Text",
          "value": "
",
        },
      ],
      "closingElement": ClosingElementNode {
        "end": 108,
        "loc": Object {
          "end": Object {
            "column": 7,
            "line": 8,
          },
          "start": Object {
            "column": 0,
            "line": 8,
          },
        },
        "range": Array [
          101,
          108,
        ],
        "start": 101,
        "type": "ClosingElement",
      },
      "end": 108,
      "loc": Object {
        "end": Object {
          "column": 7,
          "line": 8,
        },
        "start": Object {
          "column": 0,
          "line": 2,
        },
      },
      "openingElement": OpeningElementNode {
        "attributes": Array [],
        "end": 22,
        "loc": Object {
          "end": Object {
            "column": 6,
            "line": 2,
          },
          "start": Object {
            "column": 0,
            "line": 2,
          },
        },
        "range": Array [
          16,
          22,
        ],
        "start": 16,
        "type": "OpeningElement",
      },
      "range": Array [
        16,
        108,
      ],
      "selfClosing": false,
      "start": 16,
      "tagName": "html",
      "type": "Element",
    },
    TextNode {
      "end": 109,
      "loc": Object {
        "end": Object {
          "column": 8,
          "line": 8,
        },
        "start": Object {
          "column": 7,
          "line": 8,
        },
      },
      "range": Array [
        108,
        109,
      ],
      "start": 108,
      "type": "Text",
      "value": "
",
    },
  ],
  "comments": Array [
    CommentNode {
      "end": 48,
      "loc": Object {
        "end": Object {
          "column": 16,
          "line": 4,
        },
        "start": Object {
          "column": 4,
          "line": 4,
        },
      },
      "range": Array [
        36,
        48,
      ],
      "start": 36,
      "type": "Comment",
      "value": " foo ",
    },
  ],
  "end": 109,
  "loc": Object {
    "end": Object {
      "column": 8,
      "line": 8,
    },
    "start": Object {
      "column": 0,
      "line": 1,
    },
  },
  "range": Array [
    0,
    109,
  ],
  "start": 0,
  "tokens": Array [],
  "type": "Root",
}
`;
